### ISSUE TITLE
Fork fasd

### ISSUE THREAD

**Issue by josephholsten** [shell-fishes/meta#2]
Per https://github.com/clvv/fasd/issues/113, @clvv is in support of a community fork of fasd. It's not clear if we should use the name `fasd` for our fork or if we should change it. It's also not clear which existing fork we should descend from, but https://github.com/infokiller/fasd seems like a good start.

We will also want to attempt to review the 27 open pull requests and determine if any should be merged.

**Comment by FreddieOliveira** [comment]
I agree about taking @infokiller fork as our base. And let's not only review the pull requests, but also the opened issues.

**Comment by infokiller** [comment]
In that case I vote to create the fork so we can start collaborating on it. Since we're not sure how to name it yet (and if @clvv is cool with using fasd), we can just use a temporary name (like `fasd-ng`), add a warning that this is a WIP repo that will be renamed soon, and then at some point [rename it](https://docs.github.com/en/free-pro-team@latest/github/administering-a-repository/renaming-a-repository). Even after the renaming, the old git URLs should still work (though Github recommends updating them anyway).

**Comment by FreddieOliveira** [comment]
I'm cool with that. We could use nfasd as temporary name, short for neo-fasd.

**Comment by clvv** [comment]
If backwards compatibility is not broken, I'm not opposed the the fork superseding the old `fasd`.

Taking a brief-look at the fork being suggested, it seems that it has diverged from the original philosophy. `fasd` was written in plain POSIX-compliant shell (no `bash`-particular features). Restricting to POSIX-compliant shell features makes it so that there are less bugs (the POSIX-compliant part of shells and other tools don't usually change). But it makes the code very much unreadable.

Overall, here are some of my recommendations for the first steps of the fork:
1. What is the goal of the fork? (adding features? fixing bugs? improving performance?)
1. What coding guideline do we impose? (restrict to POSIX or allow bash features?)
1. How do we plan to integrate with different desktop platforms? (I suggest having separate maintainers when necessary for each desktop/os backend)

Looking back, using POSIX-shell is very restricting. The core matching capability and performance is very much bottlenecked (we are required to use shell, sed, or awk to do this, and calling external binaries is very inefficient). The the core, `fasd` is just a path database with a fuzzy matching interface, with supporting shell scripts for each shell and backend. The core is probably better written in C or Rust.

**Comment by infokiller** [comment]
These are very good points @clvv 
For the fuzzy matching part, I opted to use fzf, which I think is better than using sed/awk: more readable, more performant, and more feature-full (has an interactive UI, has a preview window, and other stuff). The disadvantage is that it requires another dependency, while sed/awk are almost always installed. But even then I think using fzf here is a no-brainer.
One thing that could be improved is that fzf doesn't consider "frecency" when sorting matches. Possibly, one could use fzf as a golang library and add it as well, but I didn't look into it.

Other operations like adding/updating fasd with a path or listing paths could probably also be done more easily and efficiently in a "real" programming language.

**Comment by FreddieOliveira** [comment]
Maybe we could write one script per supported shell with code related only to that shell using its own coding conventions, ie. zsh script uses zsh conventions; bash script uses bash; fish uses fish; etc. While the core part we write in C or whatever. This solves this POSIX compliance problem and boosts performance.

As per retrocompatibility I think it's a must, focusing first in fixing existing bugs and then, very carefully, adding new functionalities.

**Comment by clarity20** [comment]
It's cool that you all are interested in making _fasd_ even better.

I created [fasder](https://github.com/clarity20/fasder) a few years back specifically to speed up [performance on Cygwin](https://github.com/clvv/fasd/pull/97). I squeezed out some nice performance gains but introduced a series of bashisms at the same time.

I'm interested to see if this discussion bears fruit and also in how I might help out in that process.

Cheers!

**Comment by ralphee** [comment]
any movement ?

**Comment by marko-radosavljevic** [comment]
I'm also interested in the future of this project.

I think someone should step up and take the lead, because there are many decisions to be made. Questions from @clvv are a good start.

Instead of checking all the forks, maybe it's easier to just fork the original repo, and start from there. This way we avoid major design decisions. There are many open PRs on the original that we can merge, and other fork owners can contribute here with their progress.

In any case, whoever has the necessary permissions should fork either original or some other fork of their choice, so people can start contributing. ^^

**Comment by infokiller** [comment]
I personally don't want to invest too much into my fork before a proper review and comparison is done with alternatives such as:

- https://github.com/gsamokovarov/jump
- https://github.com/euank/pazi
- https://github.com/ajeetdsouza/zoxide

All of these use "proper" programming languages and can build static binaries for multiple platforms, meaning that they're probably faster and easier to maintain than fasd.
We need to be sure of the value that fasd brings with all these existing alternatives (which didn't exist when fasd was originally created). Otherwise, I feel my energy is better spent in supporting the other projects.

**Comment by marko-radosavljevic** [comment]
Yeah, I agree with you, @infokiller.

Thanks for the links too, I never heard for [pazi](https://github.com/euank/pazi), for example. 

Pazi benchmarks:
https://github.com/euank/pazi/blob/master/docs/Benchmarks.md

Benchmarks including [jump](https://github.com/gsamokovarov/jump):
https://github.com/euank/pazi/pull/83

They are all pretty fast tho, I would be mainly interested in feature comparison.

Despite the modern tooling, there are still many _fasd_ users, having an active community would provide some value. If nothing else, some basic maintenance.

**Comment by laur89** [comment]
Had a quick look around the listed alternatives, and looks like they all are missing one major feature of fasd - managing files & dirs separately. That allows creating separate functions for opening a file or cd-ing to a dir:

	e() { 
		local file;
		file="$(fasd -Rfl "$@" | fzf -1 -0 --no-sort +m --exit-0)";
		[[ -f "$file" ]] && $EDITOR -- "$file" && return 0 || return 1
	}

	d() {
		local dir;
		dir="$(fasd -Rdl "$@" | fzf -1 -0 --no-sort +m --exit-0)";
		[[ -d "$dir" ]] && cd -- "$dir" && return 0 || return 1
	}

**Comment by infokiller** [comment]
@laur89 maybe not out of the box, but it's probably easy to pipe their output into a something that filters what you need (something like `your_fasd_alternative | find {} -mindepth 1 -maxdepth 1 -type f`.

**Comment by benmccann** [comment]
@laur89 here's a discussion in the pazi repo about some of the complications around implementing that feature: https://github.com/euank/pazi/issues/4
